exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go

```diff
@@ -110,52 +125,101 @@ func (r *ETCDMachineSnapshotReconciler) Reconcile(ctx context.Context, req ctrl.
 	}
 
 	// Ensure the finalizer is present
-	if !controllerutil.ContainsFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer) {
-		controllerutil.AddFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer)
-		if err := r.Client.Update(ctx, etcdMachineSnapshot); err != nil {
-			log.Error(err, "Failed to add finalizer to EtcdMachineSnapshot")
-			return ctrl.Result{}, err
-		}
-	}
+	controllerutil.AddFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer)
 
 	return r.reconcileNormal(ctx, etcdMachineSnapshot)
 }
 
+func (r *ETCDMachineSnapshotReconciler) newScope(ctx context.Context, etcdMachineSnapshot *snapshotrestorev1.ETCDMachineSnapshot) (*snapshotScope, error) {
+	// Get the cluster object.
+	cluster := &clusterv1.Cluster{}
+
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: etcdMachineSnapshot.Namespace, Name: etcdMachineSnapshot.Spec.ClusterName}, cluster); err != nil {
+		return nil, fmt.Errorf("failed to get cluster: %w", err)
+	}
+
+	machines, err := collections.GetFilteredMachinesForCluster(ctx, r.Client, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("failed to collect machines for cluster: %w", err)
+	}
+
+	controlPlaneMachines := machines.Filter(collections.ControlPlaneMachines(cluster.Name))
+	targetMachineCandidates := controlPlaneMachines.Filter(func(machine *clusterv1.Machine) bool {
+		return machine.Name == etcdMachineSnapshot.Spec.MachineName
+	}).UnsortedList()
+
+	if len(targetMachineCandidates) < 1 {
```

Comment:
should it be `!=` ?

Reply:
It makes little difference, current assumption is that the first in the list should be used.